### PR TITLE
C++ implementation of GetBestRMS and GetConformerRMS

### PR DESCRIPTION
**Pull request by gedeck**
This is not a PR, but a request for comments. The Python tests still work, however the implementation is incomplete. I'll add a few comments where I have specific questions.

### CODE REVIEW

File: Code/GraphMol/MolAlign/AlignMetrics.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+namespace MolAlign {
+
+double getConformerRMS(ROMol &mol, unsigned int confId1, unsigned int confId2,
+                       const std::vector<unsigned int> *atomIds,
+                       bool prealigned) {
+  if (!prealigned) {
+    static const unsigned int confIds[] = {confId1, confId2};
+    const std::vector<unsigned int> confIds2(
+        confIds, confIds + sizeof(confIds) / sizeof(confIds[0]));
+    MolAlign::alignMolConformers(mol, atomIds, &confIds2);
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/AlignMetrics.cpp]: Is there a better way to create the `std::vector<unsigned int>` for the two confIDs?

File: Code/GraphMol/MolAlign/AlignMetrics.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  ssr /= nAtoms;
+  return sqrt(ssr);
+}
+
+double getBestRMS(const ROMol &refMol, ROMol &prbMol, int refCid, int prbCid,
+                  std::vector<MatchVectType> *atomMaps) {
+  RDGeom::Point3DConstPtrVect refPoints, prbPoints;
+
+  if (!atomMaps || atomMaps->size() == 0) {
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/AlignMetrics.cpp]: I have the second condition here, because of the way I solved the Python integration where I create an empty `std::vector<MatchVectType>` as default for missing atomMaps.

File: Code/GraphMol/MolAlign/AlignMetrics.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                    to all atoms)
+  \param prealigned by default the conformers are assumed to be unaligned and
+                    will therefore be aligned to the first conformer
+
+  <b>Returns</b>
+  RMSD value
+*/
+double getConformerRMS(ROMol &mol, unsigned int confId1, unsigned int confId2,
+                       const std::vector<unsigned int> *atomIds = 0,
+                       bool prealigned = false);
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/AlignMetrics.h]: `0` or `NULL`?

File: Code/GraphMol/MolAlign/CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,7 @@
-rdkit_library(MolAlign AlignMolecules.cpp O3AAlignMolecules.cpp
+rdkit_library(MolAlign AlignMolecules.cpp O3AAlignMolecules.cpp AlignMetrics.cpp
               LINK_LIBRARIES MolTransforms SubstructMatch Alignment)
 
-rdkit_headers(AlignMolecules.h O3AAlignMolecules.h DEST GraphMol/MolAlign)
+rdkit_headers(AlignMolecules.h AlignMetrics.h O3AAlignMolecules.h DEST GraphMol/MolAlign)
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/CMakeLists.txt]: Are these all the changes needed?

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp [hunk 5]
```diff
@@ -197,6 +198,39 @@ double AlignMolecule(ROMol &prbMol, const ROMol &refMol, int prbCid = -1,
...
+
+double getBestRMS(const ROMol &ref, ROMol &probe, int refConfId,
+                  int probeConfId, python::object maps = python::list()) {
+  double rmsd;
+  if (maps) {
+    std::vector<MatchVectType> atomMaps;
+    //    BOOST_FOREACH (const python::object &map, *maps) {
+    //      MatchVectType *atomMap = _translateAtomMap(map);
+    //      atomMaps.push_back(*atomMap);
+    //    }
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: How do I convert the maps (list of atom mappings = list of list) to a vector of atomMaps? What I have commented here doesn't work.

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp [hunk 6]
```diff
@@ -197,6 +198,39 @@ double AlignMolecule(ROMol &prbMol, const ROMol &refMol, int prbCid = -1,
...
+    rmsd = 3.14;
+  } else {
+    std::vector<MatchVectType> atomMaps;
+    rmsd = MolAlign::getBestRMS(ref, probe, refConfId, probeConfId, &atomMaps);
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: I tried:
```
MolAlign::getBestRMS(ref, probe, refConfId, probeConfId)
```
under the assumption that atomMaps would be then be set to 0, however this didn't work. Any idea why?

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp [hunk 7]
```diff
@@ -661,6 +695,59 @@ BOOST_PYTHON_MODULE(rdMolAlign) {
...
+      - atomIds     (optional) list of atom ids to use a points for\n\
+                    alingment - defaults to all atoms\n\
+      - prealigned  (optional) by default the conformers are assumed\n\
+                    be unaligned and will therefore be aligned to the\n\
+                    first conformer\n\
+     \n\
+     RETURNS\n\
+     RMSD value\n\
+     \n";
+  python::def("GetConformerRMS", RDKit::getConformerRMS,
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: Is the namespace RDKit reserved for the python-C++ interface?

File: Code/GraphMol/MolAlign/testMolAlign.cpp [hunk 8]
```diff
@@ -384,8 +398,9 @@ void testMMFFO3AConstraints() {
   TEST_ASSERT(o3a);
   o3a->align();
   delete o3a;
-  double d = (m2.getConformer().getAtomPos(cIdx) -
-              m1->getConformer().getAtomPos(cIdx)).length();
+  double d =
+      (m2.getConformer().getAtomPos(cIdx) - m1->getConformer().getAtomPos(cIdx))
```
Comment (gedeck) [inline Code/GraphMol/MolAlign/testMolAlign.cpp]: Did you format this file? If yes, my linewidth is probably set wrongly. What do you use?

File: Code/GraphMol/MolAlign/AlignMetrics.h (hunk 3 above)
Comment (greglandrum) [inline Code/GraphMol/MolAlign/AlignMetrics.h]: NULL please

File: Code/GraphMol/MolAlign/CMakeLists.txt (hunk 4 above)
Comment (greglandrum) [inline Code/GraphMol/MolAlign/CMakeLists.txt]: yep, that's it

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp (hunk 5 above)
Comment (greglandrum) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: I always have to think about this/look it up.
In the interests of getting this review done in a timely manner, I'm going to punt on this for the moment. I will come back to it.

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp (hunk 6 above)
Comment (greglandrum) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: Should have been. What happened instead?

As an aside: If you want to be explicit about it,  you can just directly pass a NULL. I don't really care for creating an empty object and then passing that.

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp [hunk 9]
```diff
@@ -661,6 +695,59 @@ BOOST_PYTHON_MODULE(rdMolAlign) {
       docString.c_str());
 
   docString =
+      "Returns the RMS between two conformations.\n\
+     \n\
+     By default, the conformers will be aligned to the first conformer\n\
+     of the molecule (i.e. the reference) before RMS calculation and,\n\
+     as a side-effect, will be left in the aligned state.\n\
```
Comment (greglandrum) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: I think you mean here: "as a side-effect the conformers will be left in aligned state"

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp (hunk 7 above)
Comment (greglandrum) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: no, it's all over the place. Normally when you are writing RDKit code on the C++ side you are inside a `namespace RDKit` block, so you don't notice. It shows up in the C++ tests a lot though.

File: Code/GraphMol/MolAlign/testMolAlign.cpp (hunk 8 above)
Comment (greglandrum) [inline Code/GraphMol/MolAlign/testMolAlign.cpp]: I format the files using clang-format. There is a setting file for it at the root of the rdkit tree. If you are going to use an automatic formatter, it should be this one. Otherwise please do not automatically format C++ code.
And, yes, I really need to finish the Contributing document that explains all this stuff. :-S

File: Code/GraphMol/MolAlign/testMolAlign.cpp [hunk 10]
```diff
@@ -879,7 +897,11 @@ int main() {
   std::cout << "***********************************************************\n";
   std::cout << "Testing MolAlign\n";
 
-#if 1
+  std::cout << "\t---------------------------------\n";
+  std::cout << "\t test1AlignMetrics \n\n";
+  test1AlignMetrics();
+
+#if 0
```
Comment (greglandrum) [inline Code/GraphMol/MolAlign/testMolAlign.cpp]: Please don't check in code that has tests that have been disabled like this. It's way too easy to forget to re-enable them later.

File: rdkit/Chem/PeriodicTable.py [hunk 11]
```diff
@@ -15,106 +15,105 @@
 
 """
```
Comment (greglandrum) [inline rdkit/Chem/PeriodicTable.py]: I don't think that the changes to this file are part of this PR

File: rdkit/Chem/TorsionFingerprints.py [hunk 12]
```diff
@@ -9,56 +9,46 @@
     According to a paper from Schulz-Gasch et al., JCIM, 52, 1499-1512 (2012).
```
Comment (greglandrum) [inline rdkit/Chem/TorsionFingerprints.py]: These changes also don't belong in this PR

File: rdkit/Chem/UnitTestCatalog.py [hunk 13]
```diff
@@ -14,7 +14,7 @@
 from rdkit import RDConfig
```
Comment (greglandrum) [inline rdkit/Chem/UnitTestCatalog.py]: more changes for another PR

File: rdkit/Chem/UnitTestMol3D.py [hunk 14]
```diff
@@ -3,18 +3,47 @@
 """unit testing code for 3D stuff
```
Comment (greglandrum) [inline rdkit/Chem/UnitTestMol3D.py]: and some more for another PR

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp (hunk 7 above)
Comment (gedeck) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: So, the molecule alignment Python interfaces could all just be in the MolAlign namespace for consistency.

File: Code/GraphMol/MolAlign/testMolAlign.cpp (hunk 8 above)
Comment (gedeck) [inline Code/GraphMol/MolAlign/testMolAlign.cpp]: This is what I use:
```
clang-format version 3.9.0 (tags/RELEASE_390/final)
```
it uses the .clang-format file from RDkit, I checked this.

Could be a change in the formatter. I've seen this in yapf as well.

File: rdkit/Chem/UnitTestMol3D.py (hunk 14 above)
Comment (gedeck) [inline rdkit/Chem/UnitTestMol3D.py]: This contains the tests for the GetBestRMS and GetConformerRMS, so in this respect it belongs to the PR. The changes are however for the TorsionFingerprint (as the branch is named wrongly).

File: Code/GraphMol/MolAlign/AlignMetrics.cpp [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    // we have to figure out all mappings between the two molecule
+    const bool uniquify = true;
```
Comment (bp-kelley) [inline Code/GraphMol/MolAlign/AlignMetrics.cpp]: I think I added this comment in a review for DoliathGavid, greg knows my penchant for taking a stiff drink when reading function invocations like:

```
foo(2, false, true, false, false);
```

File: Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp [hunk 16]
```diff
@@ -197,6 +198,39 @@ double AlignMolecule(ROMol &prbMol, const ROMol &refMol, int prbCid = -1,
...
+  return rmsd;
+}
+
+double getBestRMS(const ROMol &ref, ROMol &probe, int refConfId,
+                  int probeConfId, python::object maps = python::list()) {
+  double rmsd;
+  if (maps) {
+    std::vector<MatchVectType> atomMaps;
+    //    BOOST_FOREACH (const python::object &map, *maps) {
+    //      MatchVectType *atomMap = _translateAtomMap(map);
```
Comment (bp-kelley) [inline Code/GraphMol/MolAlign/Wrap/rdMolAlign.cpp]: You'll need to use the magic of python::extract.  You'll probably also need to make tuple and list versions, I don't know how to do sequences yet.

It will be something like:

```
template<class T, class U>
std::vector<std::vector<U>> ConvertVectVect(T vectvect) {
  std::vector<std::vector<U> > vect;
  size_t size = python::extract<unsigned int>(vectvect.attr("__len__")());
  vect.resize(size);
  for(size_t i=0; i<size; ++i) {
    unsigned int len1 = python::extract<unsigned int>(vectvect[i].attr("__len__")());
    std::vector<U> &v = vect[i];
    v.reserve(len1);
    for(unsigned int j=0;j<len1;++j){
      U u = python::extract<U>(vectvect[i][j]);
      v.push_back(u);
    }
  }
  return vect;
}
```

with usage here:

```
double getBestRMS(const ROMol &ref, ROMol &probe, int refConfId,
                                  int probeConfId, python::list maps) {
...
std::vector<MatchVectType> atomMaps = ConvertVectVec<python::list, MatchType>(maps);
...
}

double getBestRMS(const ROMol &ref, ROMol &probe, int refConfId,
                                  int probeConfId, python::tuple maps) {
...
std::vector<MatchVectType> atomMaps = ConvertVectVec<python::tuple, MatchType>(maps);
...
}
```

I would put initialize the default args in the wrapped arguments.
